rosidl_generator_type_hash/rosidl_generator_type_hash/__init__.py

```diff
@@ -0,0 +1,320 @@
+            ):
+                included_types.append(member.type.value_type)
+
+        self.includes = [
+            str(Path(*t.namespaced_name()).with_suffix('.in.json'))
+            for t in included_types
+        ]
+
+        self.rel_path = Path(*self.namespaced_type.namespaced_name()[1:])
+        self.include_path = Path(*self.namespaced_type.namespaced_name())
+
+        self.json_in = {
+            'type_description': self.individual_type_description,
+            'includes': self.includes,
+        }
+
+    def write_json_in(self, output_dir) -> List[str]:
+        """Return list of written files."""
+        generated_files = []
+        for key, val in self.subinterfaces.items():
+            generated_files += val.write_json_in(output_dir)
+
+        json_path = output_dir / self.rel_path.with_suffix('.in.json')
+        json_path.parent.mkdir(parents=True, exist_ok=True)
+        with json_path.open('w', encoding='utf-8') as json_file:
+            json_file.write(json.dumps(self.json_in, indent=2))
+        return generated_files + [str(json_path)]
+
+    def write_json_out(self, output_dir: Path, includes_map: dict) -> List[str]:
+        """Return list of written files."""
+        generated_files = []
+        for key, val in self.subinterfaces.items():
+            generated_files += val.write_json_out(output_dir, includes_map)
+
+        # Recursively load includes from all included type descriptions
+        pending_includes = self.includes[:]
+        loaded_includes = {}
+        while pending_includes:
+            process_include = pending_includes.pop()
+            if process_include in loaded_includes:
+                continue
+            p_path = Path(process_include)
+            assert(not p_path.is_absolute())
+            include_package = p_path.parts[0]
+            include_file = includes_map[include_package] / p_path.relative_to(include_package)
+
+            with include_file.open('r') as include_file:
+                include_json = json.load(include_file)
+
+            loaded_includes[process_include] = include_json['type_description']
+            pending_includes.extend(include_json['includes'])
+
+        # Sort included type descriptions alphabetically by type name
+        self.json_out = {
+            'type_description': self.json_in['type_description'],
+            'referenced_type_descriptions': sorted(
+                loaded_includes.values(), key=lambda td: td['type_name'])
+        }
+
+        json_path = output_dir / self.rel_path.with_suffix('.json')
+        with json_path.open('w', encoding='utf-8') as json_file:
+            json_file.write(json.dumps(self.json_out))  # NOTE: no whitespace!
+        return generated_files + [str(json_path)]
+
+    def _calculate_hash_tree(self) -> dict:
+        prefix = f'RIHS{RIHS_VERSION}_'
+        json_out_repr = json.dumps(self.json_out)  # NOTE: no whitespace!
```

Comment:
might be worth specifying all the formatting args here explicitly?

Reply:
agreed, done